### PR TITLE
osd: Extreme conditions cause ignore peer missing

### PR DESCRIPTION
**Pull request by Aran85**
the analysis of circumstance:

<epoch 16259 [23,13,12,5]/p23>
calc_acting osd.5 14.1a( v 16253'3749 (7924'2167,16253'3749] local-lis/les=16255/16256 n=1 ec=2558/2558 lis/c 16255/16248 les/c/f 16256/16249/0 16259/16259/16259)
calc_acting osd.12 14.1a( v 16253'3749 (7924'2167,16253'3749] local-lis/les=16255/16256 n=1 ec=2558/2558 lis/c 16255/16248 les/c/f 16256/16249/0 16259/16259/16259)
calc_acting osd.13 14.1a( v 16253'3749 (7924'2167,16253'3749] local-lis/les=16255/16256 n=1 ec=2558/2558 lis/c 16255/16248 les/c/f 16256/16249/0 16259/16259/16259)
calc_acting osd.23 14.1a( v 16253'3750 (7924'2167,16253'3750] local-lis/les=16248/16249 n=0 ec=2558/2558 lis/c 16255/16248 les/c/f 16256/16249/0 16259/16259/16259)

when osd.23 do peering as primary, he choose osd.5 as auth log(because his les is smaller).
so he rewind 16253'3750, and mark the obj as missing. then he notify the monitor to update up_thru, and persist his pg_info.
then new osdmap change up set, new peering begin.

<epoch  16260 [26,23,13,12,5]/p26>
calc_acting osd.5 14.1a( v 16253'3749 (7924'2167,16253'3749] local-lis/les=16255/16256 n=1 ec=2558/2558 lis/c 16255/16248 les/c/f 16256/16249/0 16259/16259/16259)
calc_acting osd.12 14.1a( v 16253'3749 (7924'2167,16253'3749] local-lis/les=16255/16256 n=1 ec=2558/2558 lis/c 16255/16248 les/c/f 16256/16249/0 16259/16259/16259)
calc_acting osd.13 14.1a( v 16253'3749 (7924'2167,16253'3749] local-lis/les=16255/16256 n=1 ec=2558/2558 lis/c 16255/16248 les/c/f 16256/16249/0 16259/16259/16259)
calc_acting osd.23 14.1a( v 16253'3750 (7924'2167,16253'3750] local-lis/les=16248/16249 n=0 ec=2558/2558 lis/c 16255/16248 les/c/f 16256/16249/0 16259/16259/16259)
calc_acting osd.26 14.1a( v 16253'3750 (7924'2167,16253'3750] local-lis/les=16248/16249 n=0 ec=2558/2558 lis/c 16255/16248 les/c/f 16256/16249/0 16259/16259/16259)

now osd.26 choose osd.5 as auth log again, and rewind 16253'3750, but in get_missing stage,
he think osd.23's info identical to his ,so skip get_log from osd.23, and mark osd.23 *has no missing*(wrong place).

when pg go active and write the missing object, rep pg will hit "assert(!parent->get_log().get_missing().is_missing(soid))" in ReplicatedBackend::do_repop()

tracker: https://tracker.ceph.com/issues/23031

Signed-off-by: Zengran Zhang <zhangzengran@sangfor.com.cn>

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by Aran85** [comment]
@jdurgin @neha-ojha need help to review, thanks!

File: src/osd/PG.cc [hunk 1]
```diff
@@ -5859,6 +5859,9 @@ void PG::fulfill_info(
   // info
   dout(10) << "sending info" << dendl;
   notify_info = make_pair(from, info);
+  if (!pg_log.get_missing().num_missing()) {
+    notify_info.second.have_missing = true;
```
Comment (jdurgin) [inline src/osd/PG.cc]: abbreviate to ```have_missing = pg_log.get_missing().have_missing()```

File: src/osd/PG.cc [hunk 2]
```diff
@@ -9282,7 +9285,8 @@ PG::RecoveryState::GetMissing::GetMissing(my_context ctx)
     }
 
     if (pi.last_update == pi.last_complete &&  // peer has no missing
-	pi.last_update == pg->info.last_update) {  // peer is up to date
+	pi.last_update == pg->info.last_update &&  // peer is up to date
+	pi.have_missing == false) {
```
Comment (jdurgin) [inline src/osd/PG.cc]: ```!pi.have_missing``` is more usual style for comparing bools

File: src/osd/osd_types.h [hunk 3]
```diff
@@ -2480,6 +2480,8 @@ struct pg_info_t {
   pg_history_t history;
   pg_hit_set_history_t hit_set;
 
+  bool have_missing;
```
Comment (jdurgin) [inline src/osd/osd_types.h]: should add this to ```operator==``` and ```operator<<```for completeness.

```encode()``` and ```decode()``` also need this added so it is actually sent across the wire

Comment (jdurgin) [reply src/osd/osd_types.h]: ```have_missing``` would only be potentially true for ```pg_info_t``` sent to the primary, and stored in ```peer_info```. It's only used before the next step of peering when we fetch the actual log and missing set from the secondaries. Once we have actual missing set, ```peer_info[peer]->have_missing``` shouldn't be read again.

These ```peer_info``` will be reset whenever peering restarts, and I think all that is needed is to make ```PG::proc_replica_info()``` update ```peer_info[peer]->have_missing``` even when ```last_update``` matches the  ```pg_info_t``` we already have for that peer.

Comment (Aran85) [reply src/osd/osd_types.h]: thanks the guidance, updated

**Comment by Aran85** [comment]
@jdurgin I'm glade to add some test to qa, but the problem is complex, we may need to add some inject point  to the peering code, which will make the code not so clean. Even though it is also hard to  accurately control the several round of peering amount several osds. any suggestions? thanks!

**Comment by jdurgin** [comment]
Using flags like noout, norecover, nobackfill can give you some control. You can stop a specific osd, do a write to the rest, then start the osd again, perhaps using upmap to make the osds involved in the pg your choice.

If we need to add some injection to peering for testing this, by all means go ahead. Better to have tested this path to avoid regression.

**Comment by Aran85** [comment]
this the steps could reproduct same "assert"



\# prepare 
MON=1 OSD=3 MDS=0 RGW=0 MGR=1 OCS=0 ../src/vstart.sh --new -x --localhost --bluestore --debug
./bin/ceph osd pool create rep 1 1
./bin/ceph osd pool set rep min_size 2
./bin/rados -p rep put test /etc/hosts

\# here i add a new option called "osd_assert_do_repop", make it true to let non-primary osd assert when do_reopop()
./bin/ceph daemon osd.2 config set osd_assert_do_repop true

\# let osd.2 assert here
./bin/rados -p rep put test /etc/hosts

\# next steps will let osd.0 mark the object as missing and keep last_complete == last_update
./bin/init-ceph stop osd.1
./bin/ceph daemon osd.0 injectdataerr rep test
./bin/ceph daemon osd.0 config set bluestore_debug_inject_read_err true
./bin/init-ceph start osd.2

\# ... wait for osd.0 active

\# stop osd.0 temporarily 
./bin/init-ceph stop osd.0
./bin/init-ceph stop osd.2

\# next steps will let osd.1 mark the object as and keep last_complete == last_update
./bin/init-ceph start osd.1
./bin/ceph daemon osd.1 injectdataerr rep test
./bin/ceph daemon osd.1 config set bluestore_debug_inject_read_err true
./bin/init-ceph start osd.2

\# ... wait for osd.1 active

./bin/init-ceph stop osd.2

\# now 
\# the osd.1's pglog is: pg[1.0( v 15'2 lc 14'1 (0'0,15'2] m=1  --- will be primary next
\# the osd.0's pglog is: pg[1.0( v 15'2  (0'0,15'2]   m=1
./bin/init-ceph start osd.0

\# osd.1 will think osd.0 do not missing the object, and recovery the object..
\# ... wait for active and recovery

./bin/rados -p rep put test /etc/hosts  ==> assert(!parent->get_log().get_missing().is_missing(soid))

# summary
on the above steps, the key point is make the non-primary's pglog satisfy 3 requirements:
1. has missing
2. last_complete == last_update ***
3. last_update == primary->info.last_update

the "read error injection" is the recipe to reach the requirement 2, i have no ideas if there has other recipes..

@jdurgin  do the test steps acceptable?

**Comment by liewegas** [comment]
@Aran85 Just noticed this bug.  The fix looks good to me, too!  The easiest way to construct the test is to use the qa/standalone/* framework.  Look at, for instance, ``qa/standalone/osd/pg-split-merge.sh`` for an example.  Individual tests are quick to run from you build dir with ``../qa/run-standalone.sh osd/pg-split-merge.sh TEST_foo``.

**Comment by liewegas** [comment]
something like (untested):
```
#!/usr/bin/env bash
source $CEPH_ROOT/qa/standalone/ceph-helpers.sh

function run() {
    local dir=$1
    shift

    export CEPH_MON="127.0.0.1:7155" # git grep '\<7155\>' : there must be only one
    export CEPH_ARGS
    CEPH_ARGS+="--fsid=$(uuidgen) --auth-supported=none "
    CEPH_ARGS+="--mon-host=$CEPH_MON"

    local funcs=${@:-$(set | sed -n -e 's/^\(TEST_[0-9a-z_]*\) .*/\1/p')}
    for func in $funcs ; do
        $func $dir || return 1
    done
}

function TEST_peer_missing() {
    local dir=$1

    setup $dir || return 1
    run_mon $dir a --osd_pool_default_size=3 || return 1
    run_mgr $dir x || return 1
    run_osd $dir 0 || return 1
    run_osd $dir 1 || return 1
    run_osd $dir 2 || return 1

    ceph osd pool create rep 1 || return 1
    ceph osd pool set min_size rep 2 || return 1
    rados -p rep put test /etc/hosts || return 1

    ceph daemon osd.2 config set osd_assert_do_repop true || return 1

    rados -p rep put test /etc/hosts || return 1

    kill_daemons $dir TERM osd.1 || return 1
    ceph daemon osd.0 injectadataerr rep test || return 1
    ceph daemon osd.0 config set bluestore_debug_inject_read_err true || return 1
    activate_osd $dir 2 || return 1

    sleep 10 # wait for pgs to go active

    kill_daemons $dir KILL osd.0 || return 1
    kill_daemons $dir KILL osd.2 || return 1

    activate_osd $dir 1 || return 1
    ceph daemon osd.1 injectdataerr rep test || return 1
    ceph daemon osd.1 config set bluestore_debug_inject_read_err true || return 1

    activate_osd $dir 2 || return 1

    sleep 10

    kill_daemons $dir KILL osd.2 || return 1

    activate_osd $dir 0 || return 1

    sleep 10

    rados -p rep put test /etc/hosts || return 1

    wait_for_clean || return 1
}



main osd-peering "$@"

# Local Variables:
# compile-command: "cd ../.. ; make -j4 && test/osd/pg-split-merge.sh"
# End:
```

**Comment by Aran85** [comment]
thanks so much, I will try.

**Comment by Aran85** [comment]
@liewegas @jdurgin  a littel change, I found we need make send_notify() with the missing info always instead of just fulfill_info(), does the qa test acceptable?

File: src/osd/ReplicatedBackend.cc [hunk 4]
```diff
@@ -1010,6 +1010,10 @@ void ReplicatedBackend::do_repop(OpRequestRef op)
 	   << " " << m->logbl.length()
 	   << dendl;
 
+  if (cct->_conf.get_val<bool>("osd_debug_repop_crash")) {
```
Comment (liewegas) [inline src/osd/ReplicatedBackend.cc]: the get_val<> method does a lookup in a std::map, so it's slow and can't be used in any hot path.  Instead, add osd_debug_repop_crash to common/legacy_options.h and use ``cct->_conf->osd_debug_repop_crash``

Comment (Aran85) [reply src/osd/ReplicatedBackend.cc]: thanks for teaching!

**Comment by liewegas** [comment]
This looks reasonable and I think it addresses the problem, but I'm still not certain whether the explicit missing flag is the best way to approach it.  @neha-ojha ?

**Comment by neha-ojha** [comment]
@liewegas The assert in `do_repop()` described in the https://tracker.ceph.com/issues/23031 was removed in https://github.com/ceph/ceph/pull/19811/files#diff-b23db2c2a3a8221ba3ae529dfb391e0fL1013, because during async recovery(>= mimic) we can still have missing objects on async_recovery_targets, so I am bit confused how we are reproducing this failure. Or are we seeing some other failure due to the same root cause @Aran85 ?

**Comment by Aran85** [comment]
@neha-ojha  Oh, I missed the bit of async recovery on upstream.. thanks pointed it out.
1. The qa test seems more harder to verify the failure in do_repop() now, maybe need verify in objectstore transaction..  :-(
2. So the bug I think need pay more attention to, because now we remove the assert when we are even not the async_recovery_targets, that's will cause data be inconsistent silently..

**Comment by neha-ojha** [comment]
> @neha-ojha Oh, I missed the bit of async recovery on upstream.. thanks pointed it out.
> 
> 1. The qa test seems more harder to verify the failure in do_repop() now, maybe need verify in objectstore transaction..  :-(
> 2. So the bug I think need pay more attention to, because now we remove the assert when we are even not the async_recovery_targets, that's will cause data be inconsistent silently..

You can do this by checking if the transaction was applied and the object was still present in `get_local_missing()`

> 3. could you give some opinions on "whether the explicit missing flag is the best way to approach it"？

I think we should get (2) figured out first.

>    Thanks...

**Comment by Aran85** [comment]
Hi @neha-ojha  I have some doult about the patch now.. I think we should follow the *RULE* that "there will be no missing before last_complete", in other word, if pi.last_update == pi.last_complete, peer must has no missing.. but now we must have some thing wrong, so I once thought we could use the extra **pg_info.have_missing** to help find missing.

Current primary and replica will change their local info's last_complete according missing when activate()[1], So I think the above *RULE* will be true until activated, this maybe ok for replica, because primary will update his peer_info's last_complete when proc_replica_log, but the problem is the primary.Because when primary invoked out from proc_master_log, he will not update his local last_complete, so there will be an interval that violate the *RULE*, think about the peering process be interrupted before update up_thru.. 

So I think we should always add missing and change the last_complete at the same time, make sense?

[1] see https://github.com/ceph/ceph/blob/1991495a22fa74210348ffd4f261c314ef3f056c/src/osd/PeeringState.cc#L2145.

**Comment by jdurgin** [comment]
'no missing before last complete' can't be an invariant due to EIO and scrub marking arbitrary objects missing.

**Comment by jdurgin** [comment]
The other thing that makes last_complete no longer a strict boundary for missing objects is the pg log hard limit. Since the missing set is persistent, we're relying on that rather than strictly last_complete now.

I agree that pg_info.have_missing is fixing the problem and it seems like a good approach to me. Does that test case fail without your fix (but with your added asserts)? I'd like to make sure we can detect a regression in this area.

**Pull request closed without merging**